server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java

```diff
@@ -80,32 +84,38 @@ public long readLong() throws IOException {
 
     @Override
     public int readVInt() throws IOException {
-        if (slice.length - sliceIndex >= 5) {
-            final byte[] buf = slice.bytes;
-            final int offset = slice.offset;
-            byte b = buf[offset + sliceIndex++];
+        if (slice.remaining() >= 5) {
+            final byte[] buf = slice.array();
+            final int offset = slice.arrayOffset();
+            int position = slice.position();
+            byte b = buf[offset + position++];
```

Comment:
why not just use ByteBuffer::get here? (rather than operating on the underlying byte array)